x/mongo/driver/topology/server.go

```diff
@@ -540,15 +540,18 @@ func (s *Server) heartbeat(conn *connection) (description.Server, *connection) {
 			if err == nil {
 				tmpDesc := op.Result(s.address)
 				descPtr = &tmpDesc
+			} else {
+				// close the connection here rather than in the error check below to avoid calling Close on a net.Conn
+				// that wasn't successfully created
+				if conn != nil && conn.nc != nil {
```

Comment:
Seems like we should not have executed an isMaster on the connection if this condition were not true. Is the conditional really necessary?

In other words, if err == nil, then this condition should always be true, so there is no need for the conditional.

Reply:
I don't think `conn != nil` is necessary but I think the `conn.nc` check is. Iirc, for network errors, the connection type will close it's underlying `net.Conn` (I believe this was done to avoid a resource leak). That being said, I'd prefer to keep this as-is to make sure this is a low-risk change. Let me know if you disagree.

Reply:
Craig reminded me of the principle that this is violating, and the reason that it bugs me.  https://en.wikipedia.org/wiki/Law_of_Demeter, or as Craig put it "don't dot so much".  

I think the design issue could be fixed by just calling a method on `conn`, e.g. `conn.close()`. That seems like it would do the right thing, but if it feel too risky to you, I'm ok leaving it as is to keep the patch as small as possible.

Reply:
Yeah I agree with this. I'm a little hesitant because this PR is already becoming more complex than was originally intended, but I think the change you're proposing actually helps the code in terms of readability, so I added it.